### ISSUE TITLE
-Wunused: false positive for pure expressions in for-yield

### ISSUE THREAD

**Issue by hejfelix** [lampepfl/dotty#16823]
## Compiler version

3.3.0-RC2

## Minimized code

```Scala
  for
    x <- List(1,2,3) // unused pattern variable warning
    y = x * 2
  yield y
```

## Expectation

no warning since the pattern variable is used in the pure expression

**Comment by SethTisue** [comment]
same as https://github.com/scala/bug/issues/10287 perhaps? cc @som-snytt

**Comment by som-snytt** [comment]
The comment on the feature PR (at some point) was that fidelity of representation should make analysis easier in Scala 3, so hopefully this ticket is not as discouraging.

**Comment by hejfelix** [comment]
> same as https://github.com/scala/bug/issues/10287 perhaps? cc @som-snytt 

Doesn't seem to be exactly the same, but could be same root cause?

**Comment by SethTisue** [comment]
What compiler settings are needed to reproduce this? I'm not able to reproduce it with:

```
//> using scala "3.3.0-RC2"
//> using option "-Wunused:all"

object O:
  for
    x <- List(1,2,3)
    y = x * 2
  yield y
```

**Comment by som-snytt** [comment]
@SethTisue not sure how output works, but
```
➜   ./bin/scala -Wunused:privates
Welcome to Scala 3.3.1-RC1-bin-SNAPSHOT-git-8c616bf (19, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.

scala> class C { private def f = 42 }
Formatted members List()
// defined class C

scala>
➜  ./bin/scalac -Wunused:privates -d /tmp c.scala
-- Warning: c.scala:2:22 -----------------------------------------------------------------------------------------------
2 |class C { private def f = 42 }
  |                      ^
  |                      unused private member
1 warning found
```

**Comment by SethTisue** [comment]
same result (no warning) with `-Wunused:privates` (which one would hope `-Wunused:all` would include, anyway)

**Comment by som-snytt** [comment]
I intended to show that if REPL reporter doesn't report, it's conceivable that the command runner reporter doesn't report, so possibly only `scalac` reports? and maybe metals?

**Comment by hejfelix** [comment]
I tried to minimize it, but I can't seem to provoke it now. I'll see if it persists in my larger project.

**Comment by hejfelix** [comment]
It seems to happen in cross-projects:

```scala
// plugins.sbt
addSbtPlugin("org.scala-js"              % "sbt-scalajs"              % "1.13.0")
addSbtPlugin("org.portable-scala"        % "sbt-scalajs-crossproject" % "1.2.0")
addSbtPlugin("io.github.davidgregory084" % "sbt-tpolecat"             % "0.1.20")


// build.sbt

val scala3Version = "3.3.0-RC2"

lazy val root = crossProject(JSPlatform,JVMPlatform)
  .in(file("."))
  .settings(
    name := "wunused-repro",
    version := "0.1.0-SNAPSHOT",
    scalaVersion := scala3Version,
    scalacOptions ++= Seq(
      "-indent",
      "-rewrite",
      "-source",
      "future",
      "-Wunused:imports",
      "-Wunused:locals",
      "-Wunused:privates",
      "-Wunused:params",
      "-Wunused:unsafe-warn-patvars",
      "-Wunused:linted",
      "-Wunused:implicits",
    )
  )


// jvm/src/main/scala/Main.scala

@main def run = println("HEY")

val xs =
  for 
    x <- List(1,2,3) // unused pattern variable
    y = x * 2
  yield y
```

**Comment by szymon-rd** [comment]
Fixed with #17020
